File: Xpanxion Game X/system/GameElement.java
package system;


/**
 * A class representing resources, map features, and code nuggets that occupy GameSquares
 * @author Alex White
 *
 */
public class GameElement extends GameObject {
	
	private final ElementKind kind;
	
	public GameElement(ElementKind kind) {
		super(kind.getName(), kind.getElementValue());
		this.kind = kind;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + ((kind == null) ? 0 : kind.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		GameElement other = (GameElement) obj;
		if (kind != other.kind)
			return false;
		return true;
	}

}

File: csharp/xcavaxion/Game/system/GameElement.cs
using System;

namespace system {


	/**
	 * A class representing resources, map features, and code nuggets that occupy GameSquares
	 * @author Alex White
	 *
	 */
	public class GameElement : GameObject {
		
		private final ElementKind kind;
		private int volume; // cm^3
		
		/**
		 * @param kind
		 * @param volume in cm^3
		 */
		public GameElement(ElementKind kind, int volume) {
			super(kind.getName());
			this.kind = kind;
			this.volume = volume;
		}
		
		public ElementKind getKind() { return kind; }
		
		/**
		 * @param other element to be combined with this one
		 * @return {@code true} if this element successfully absorbed the given element
		 */
		public boolean absorb(GameElement other) {
			if (other.kind != this.kind)
				return false;
			volume += other.volume;
			return true;
		}

		@Override
		public int getWeight() {
			// cm^3 * (1000 g/cm^3) / 1000 = grams
			return volume * kind.getDensity() / 1000;
		}
		
		@Override
		public int getVolume() { return volume; }

		@Override
		public int getValue() {
			return kind.getValue() * getWeight();
		}

	}
}